### ISSUE TITLE
Support fetching BCP

### ISSUE THREAD

**Issue by ribose-jeffreylau** [relaton/relaton-ietf#24]
It would be good to have support for BCP.

## Symptoms

Currently, searching for BCP would fail:

```ruby
RelatonIetf::IetfBibliography.search 'BCP 47'
# => RelatonBib::RequestError: BCP.47.xml: not recognised for RFC
```

## Example source documents

- IETF BCP 47: https://tools.ietf.org/html/bcp47

**Comment by ronaldtse** [comment]
@ribose-jeffreylau agree!

**Comment by andrew2net** [comment]
@ronaldtse we get RFC documents in XML format: https://xml2rfc.tools.ietf.org/public/rfc/bibxml/reference.RFC.8341.xml
IETF BCP 47 https://tools.ietf.org/html/bcp47 is in text format. Is it possible to get BCP documets in XML format?

**Comment by ronaldtse** [comment]
The list of BCPs can be found here:
https://www.rfc-editor.org/search/rfc_search_detail.php?page=All&rfc=bcp

Unfortunately I don’t think they have XML for them, but we can find the RFC numbers of them in the list. There is also the HTML version we can fetch and scrape the necessary information from, hopefully.

**Comment by andrew2net** [comment]
HTML version contains data in a plain text. It is hard to parse a plain text.
So every BCP document has related RFC number, does it? Does it mean we can parse [RFC 4647](https://xml2rfc.tools.ietf.org/public/rfc/bibxml/reference.RFC.4647.xml) instead [BCP 47](https://tools.ietf.org/html/bcp47)?

**Comment by ronaldtse** [comment]
Good question. I think we should parse the BCP page directly.

Some BCPs are a single RFC, but some BCPs have multiple RFCs.

For example, BCP 222 is composed of two RFCs: RFC 8552 and RFC 8553.

Here's the BCP page for BCP 222: https://www.rfc-editor.org/info/bcp222

<img width="900" alt="Screen Shot 2019-07-11 at 12 44 26 AM" src="https://user-images.githubusercontent.com/11865/60987854-0a405480-a375-11e9-816d-44ccfefe9377.png">

The TXT version leads to here: https://www.rfc-editor.org/bcp/bcp222.txt

Which is actually a concatenation of RFC 8552 and RFC 8553.

The citation of BCP 222 is supposed to be this: https://www.rfc-editor.org/refs/ref-bcp222.txt

Which looks a bit wrong.

Anyway, it means that a BCP is separate from RFC, it is a collection of RFCs.

@opoudjis have we defined how we can cite a collection bibliographic entry in Relaton?

**Comment by andrew2net** [comment]
But there is also RFC 5646 related to BCP 47. Which one should we parce RFC 4647 or RFC 5646?

**Comment by ronaldtse** [comment]
>But there is also RFC 5646 related to BCP 47. Which one should we parce RFC 4647 or RFC 5646?

BCP 47 page: https://www.rfc-editor.org/info/bcp47

It indicates that BCP 47 is composed of two RFCs, 4647 and 5646.

<img width="911" alt="Screen Shot 2019-07-11 at 1 02 47 AM" src="https://user-images.githubusercontent.com/11865/60989097-9fdce380-a377-11e9-8284-3404ff73029d.png">

When you click on "Cite this BCP", it gives:

```
Phillips, A. and M. Davis, "Matching of Language Tags", BCP 47, RFC 4647, September 2006. 

Phillips, A., Ed., and M. Davis, Ed., "Tags for Identifying Languages", BCP 47, RFC 5646, September 2009.

<https://www.rfc-editor.org/info/bcp47>
```

We have to parse both RFC XML files to obtain this data.

**Comment by andrew2net** [comment]
> I think we should parse the BCP page directly.

@ronaldtse I looked through [BCP 47](https://www.rfc-editor.org/bcp/bcp47.txt), [RFC 5646](https://www.rfc-editor.org/rfc/rfc5646.txt), and [RFC 4647](https://www.rfc-editor.org/rfc/rfc4647.txt). Seems BCP 47 is just the text of RFC 5646 followed by the text of RFC 4647.
So we can fetch all needed data from https://xml2rfc.tools.ietf.org/public/rfc/bibxml/reference.RFC.5646.xml and https://xml2rfc.tools.ietf.org/public/rfc/bibxml/reference.RFC.4647.xml.
The question is: how to compose these 2 documents into one `IetfBibliographicItem` class? Most of the elements like `title` and `docid` are arrays. So we can concatenate title arrays from both RFC 5646 and RFC 4647. But `status` isn't an array.

**Comment by opoudjis** [comment]
You're not going to like the answer I give, Andrej.

BCP 47 may consist of two RFC texts, but it is itself a single reference. It is a single document, and should be cited as a single document. And citing it as a collection of documents is unhelpful, because that collection still must have a single title, and a single identifier.  And any renderer of the citation is going to look for a single title and identifier to cite it by. To use the title and identifiers of its components as titles and identifiers of their parent collection is an unacceptable abuse.

I mean it. Do not do it.

(Although the authors of BCP 47 are indeed a combination of the authors of the component documents.)

The way to cite it in Relaton is as a fairly thin document, composed of two other documents:

````
<bibitem type="standard">
<title>(get it from text)</title>
<docidentifier type="IETF">BCP 47</docidentifier>
...
<relation type="merges">
<bibitem type="standard">
<title>(RFC 5646 title)</title>
<docidentifier type="IETF">RFC 5646</docidentifier>
....
</bibitem>
</relation>
<relation type="merges">
<bibitem type="standard">
<title>(RFC 4647 title)</title>
<docidentifier type="IETF">RFC 4647</docidentifier>
....
</bibitem>
</relation>
</bibitem>
````

The relation type "merges" is one of the updates to the relaton model that we are going to propose in a forthcoming document; it isn't in the grammar yet.

There are three bibitems here. You can populate the included bibitems based on the RFC XML. You will not be able to get much out of BCP 47 text file. That's ok, get as much as you can: we'll need to work out a mapping later. BCP 47 is not a real document, so it isn't going to have a lot of real metadata to it. 

Much later. And @ronaldtse: this means a quite hairy and ad hoc mapping from relaton to RFC XML within metanorma-ietf, that I'm not going to do. :-P

I'm serious btw that I will not do it... I may write the spec for it, but the mapping of Metanorma XML to RFC XML is a huge project, that Abu Nashir has already abandoned, and that I am simply never going to get the time to do.

**Comment by andrew2net** [comment]
> <title>(get it from text)</title>

@opoudjis It seems BCP 47 doesn't have its own title. The [text](https://www.rfc-editor.org/bcp/bcp47.txt) has 2 titles from both RFCs. Should we concatenate the 2 RFCs titles in one string?

**Comment by ryzokuken** [comment]
Does BCP 47 work now? I wanted to include a reference into a draft I've been working on.

/cc @ronaldtse :innocent:

**Comment by andrew2net** [comment]
@ryzokuken Yes, it does. You can cite it by reference "IETF BCP 47"

**Comment by ryzokuken** [comment]
@andrew2net `[[[BCP47,IETF BCP 47]]]` throws an error for me `Bibliography: ERROR: No title retrieved for IETF BCP 47`.

`Gemfile.lock` says I'm on `relaton` v1.8.1 and `relaton-ietf` v1.8.0

**Comment by andrew2net** [comment]
@ryzokuken I'm not able to reproduce the error. Could you provide a document with the citation and explain how do you run the relaton script?

**Comment by ryzokuken** [comment]
@andrew2net the document is in the repo at https://github.com/ryzokuken/draft-ryzokuken-datetime-extended. It fails when I add the reference in https://github.com/ryzokuken/draft-ryzokuken-datetime-extended/blob/master/sources/sections/06-references.adoc. I run the script by running `make all` which seems to be calling `relaton` at https://github.com/ryzokuken/draft-ryzokuken-datetime-extended/blob/master/Makefile#L93-L96.

**Comment by andrew2net** [comment]
@ryzokuken could you provide a backtrace of the error? I don't see `IETF BCP 47` reference in `06-references.adoc`. Can you also provide the document with the reference added?

**Comment by ryzokuken** [comment]
@andrew2net yes, I did not commit after adding that reference at the end because the build was failing. Let me commit with the error log file and the reference added.

**Comment by ryzokuken** [comment]
@andrew2net check it out now, I just pushed to the repo. Also sending the console logs here for reference:

```
[relaton] Info: detecting backends:
Bibliography: ERROR: No title retrieved for IETF BCP 47
Bibliography: (XML Line 000326): Reference IERS is missing a document identifier (docid)
Bibliography: (XML Line 000350): Reference JAVAZDT is missing a document identifier (docid)
Crossreferences: (XML Line 000083): JAVAZDT does not have a corresponding anchor ID in the bibliography!
Crossreferences: (XML Line 000096): JAVAZDT does not have a corresponding anchor ID in the bibliography!
Crossreferences: (XML Line 000112): IERS does not have a corresponding anchor ID in the bibliography!
Metanorma XML Style Warning: (XML Line 000148): Hanging paragraph in clause
Anchors: (XML Line 000324): Anchor RFC5646 has already been used at line 324
Error: multiple instances of same ID
make: *** [Makefile:79: sources/rfc-3339.xml] Error 1
```

**Comment by andrew2net** [comment]
@ryzokuken ok, I see.
The message `Bibliography: ERROR: No title retrieved for IETF BCP 47` isn't actually an error. It is a warning and seems it comes from metanorma. In fact, the "BCP 47" is a composition of 2 documents: "RFC 4647" and "RFC 5646". The composition doesn't have its own title.
The crash is caused by document duplicating:
```
Anchors: (XML Line 000324): Anchor RFC5646 has already been used at line 324
Error: multiple instances of same ID
```
It says that document "RFC 5646" is duplicated in concatenation. Please check your sources for this duplication.

**Comment by ryzokuken** [comment]
So I cannot refer to BCP 47 (the whole) and RFC 5646 (the single RFC) in the same document?

**Comment by andrew2net** [comment]
@ryzokuken relaton doesn't have such a constraint. It seems this error is raised by metanorma. It's necessary to ask @opoudjis how to resolve the issue.

**Comment by opoudjis** [comment]
... Ah. Bugger. OK, @ryzokuken, I'm raising a separate issue about that. Metanorma should anticipate that kind of redundancy and remove them (although I'm not in love with `bibitem//bibitem/@id` being generated by relaton to begin with.)

https://github.com/metanorma/metanorma-standoc/issues/482

**Comment by ryzokuken** [comment]
Thanks a ton, @opoudjis!
